ForageMapItem endpoints

Phaete opened the pull request:
Add ForageMapItemController, ForageMapItemService and ForageMapItemRepository
Add new endpoint: /api/forageMapItems
Add create and retrieve operations for ForageMapItem 
Add JUnit and Integration tests for ForageMapItemService and ForageMapItemController

Phaete committed:
Add ForageMapItemController, ForageMapItemService and ForageMapItemRepository
Add new endpoint: /api/forageMapItems
Add create and retrieve operations for ForageMapItem
Add JUnit and Integration tests for ForageMapItemService and ForageMapItemController

Phaete committed:
Add potential fix for ForageMapItemControllerIntegrationTest failures

Phaete committed:
Another attempt for a potential fix for ForageMapItemControllerIntegrationTest failures

Phaete committed:
Fix filter conditions in ForageMapItemService
Fix JUnit tests for ForageMapItemService
Potential fix for ForageMapItemControllerIntegrationTest

Phaete committed:
Fix copy&paste error in ForageMapItemControllerIntegrationTest

sieberss reviewed (changes requested):
It is already a tough job to read PRs as long as yours.

Can you please try not to make it even more difficult with lots of lines within the code that are not relevant for the logic (tests) and complex nesting of methods (Service) ?

sieberss commented on backend/src/main/java/com/phaete/backend/forage/service/ForageMapItemService.java:
```diff
@@ -0,0 +1,87 @@
+	 * @param forageMapItemDTO the forage map item to be created
+	 * @return the newly created forage map item
+	 */
+	public ForageMapItemDTO createForageMapItem(ForageMapItemDTO forageMapItemDTO) {
+		return converterService.toDTO(
+				forageMapItemRepository.save(
+						converterService.fromDTO(forageMapItemDTO)
+				)
+		);
+	}
```
Easier to read:

```suggestion
	public ForageMapItemDTO createForageMapItem(ForageMapItemDTO forageMapItemDTO) {
	ForageMapItem newObject = converterService.fromDTO(forageMapItemDTO);
	newObject = forageMapItemRepository.save(newObject)
	return converterService.toDTO(newObject);
	}
```

sieberss commented on backend/src/main/java/com/phaete/backend/forage/service/ForageMapItemService.java:
```diff
@@ -0,0 +1,87 @@
+	 * <p>
+	 * @return a list of all forage map items that contain valid customMarkers and forageWikiItems
+	 */
+	public List<ForageMapItem> findAllForageMapItems() {
+		Map<Boolean, List<ForageMapItem>> forageMapItemMap = forageMapItemRepository.findAll()
+				.stream()
+				.collect(
+						Collectors.partitioningBy(
+								forageMapItem -> (
+										forageMapItem.getForageWikiItem() != null &&
+										forageMapItem.getCustomMarker() != null
+								)
+						)
+				);
+		return forageMapItemMap.get(true);
+	}
+
```
Why partioningBy? Wouldnt filter have the same effect? It's easier to understand.

Phaete replied:
This is in preparation for the next feature implementation.
I want to have a logger log all the "faulty" ForageMapItems to fix them. If I ignore them, the DB will eventually be full of faulty items and just take up unnecessary space

sieberss replied:
In that case I suggest an extra method that only returns the result of partitioning. Other methods can then use the appropriate Map entry.

sieberss replied:
And the check if an entry is valid, can also be extracted in an extra method.

sieberss commented on backend/src/main/java/com/phaete/backend/forage/service/ForageMapItemService.java:
```diff
@@ -0,0 +1,87 @@
+					() -> new ForageMapItemNotFoundException("Could not find forage map item with the id: " + "id")
+				)
+		);
+		if (forageMapItemDTO.forageWikiItem() == null) {
+			throw new ForageWikiItemNotFoundException("Could not find marker with the id: " + forageMapItemDTO.customMarker().id());
+		} else if (forageMapItemDTO.customMarker() == null) {
+			throw new MarkerNotFoundException("Could not find forage wiki item with the id: " + forageMapItemDTO.forageWikiItem().id());
+		}
+		return forageMapItemDTO;
```
unnecessary else makes it harder to understand

```suggestion
		if (forageMapItemDTO.forageWikiItem() == null) {
			throw new ForageWikiItemNotFoundException("Could not find marker with the id: " + forageMapItemDTO.customMarker().id());
		}
                if (forageMapItemDTO.customMarker() == null) {
			throw new MarkerNotFoundException("Could not find forage wiki item with the id: " + forageMapItemDTO.forageWikiItem().id());
		}
		return forageMapItemDTO;
```

Phaete replied:
Was a typo and supposed to be two if-statements, however there was another error in that snippet that was not caught.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/controller/ForageMapItemControllerIntegrationTest.java:
```diff
@@ -0,0 +1,413 @@
+	}
+
+	@Test
+	void findAllForageMapItems_returnsForageMapItems_withoutInvalidCustomMarker() throws Exception {
```
I don't understand what the tested method is expected to do in this test case.
Renaming might help.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/controller/ForageMapItemControllerIntegrationTest.java:
```diff
@@ -0,0 +1,413 @@
+	}
+
+	@Test
+	void findAllForageMapItems_returnsForageMapItems_withoutInvalidForageWikiItem() throws Exception {
```
I don't understand what the tested method is expected to do in this test case.
Renaming might help.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/controller/ForageMapItemControllerIntegrationTest.java:
Your XXL jsons distract from the code. Better save them in variables instead of inserting in long method chains.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/service/ForageMapItemServiceTest.java:
```diff
@@ -0,0 +1,336 @@
+	}
+
+	@Test
+	void findAllForageMapItems_expectList_WithoutInvalidWikiItem() {
```
I don't understand what the tested method is expected to do in this test case.
Renaming might help.

sieberss replied:
maybe introducing variables for what you call "Invalid Items" might also help to understand what is invalid and what should happen to invalid items.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/service/ForageMapItemServiceTest.java:
```diff
@@ -0,0 +1,336 @@
+	}
+
+	@Test
+	void findAllForageMapItems_expectList_WithoutInvalidCustomMarker() {
```
I don't understand what the tested method is expected to do in this test case.
Renaming might help.

Phaete committed:
Refactor ForageMapItemControllerIntegrationTest to extract repeated code into variables and change test names to be "easier to understand"
Fix findForageMapItemById in ForageMapItemService using and in-else statement instead of two if statements as well as their exception messages
Refactor ForageMapItemServiceTest method names to be "easier to understand"

Phaete committed:
Potential fix for ForageMapItemControllerIntegrationTest

Phaete committed:
Another potential fix for ForageMapItemControllerIntegrationTest

Phaete committed:
Fix last copy&paste error in ForageMapItemControllerIntegrationTest

sieberss commented on backend/src/test/java/com/phaete/backend/forage/controller/ForageMapItemControllerIntegrationTest.java:
```diff
@@ -0,0 +1,211 @@
+				.andExpect(content().json(forageMapItemsJsonObject));
+	}
+
+	@Test
+	void findAllForageMapItems_returnsAllForageMapItems_excludeItemsWithCustomMarkerNotInDB() throws Exception {
+		forageWikiItemRepository.save(forageWikiItem);
+		forageMapItemRepository.save(forageMapItem);
+
+		mockMvc.perform(MockMvcRequestBuilders.get("/api/forageMapItems"))
+				.andExpect(status().isOk())
+				.andExpect(content().json("[]"));
+
+	}
```
Your integration test should include both valid and invalid items. An empty list as expected could have been result of excluding too many items.

Phaete replied:
Thats what I have the other two tests for the findAll-method for. One of which tests the behaviour that valid mapItems are being shown.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/service/ForageMapItemServiceTest.java:
```diff
@@ -0,0 +1,335 @@
+		);
+		when(idService.generateId()).thenReturn("1");
+
+		ForageMapItemService forageMapItemService = new ForageMapItemService(
+				forageMapItemRepository,
+				new ConverterService(idService)
+		);
```
Why do you create new Service objects in each test instead of once for the class?

Phaete replied:
That's how I learned it. There's not really any downside afaik.

sieberss replied:
You have declared the IdService as global variable. Why not the ConverterService?

And when the ForageMapItemService is also global, you have four detracting lines less in the test.

sieberss commented on backend/src/test/java/com/phaete/backend/forage/service/ForageMapItemServiceTest.java:
```diff
@@ -0,0 +1,335 @@
+		);
+
+		when(forageMapItemRepository.findById("1")).thenReturn(
+				Optional.of(
+						new ForageMapItem(
+								"1",
+								new ForageWikiItem(
+										"1",
+										"Apple Tree",
+										ForageCategory.FRUIT,
+										ForageSource.TREE,
+										"Apple Tree",
+										ForageSeason.FALL,
+										List.of("test")
+								),
+								new CustomMarker(
+										"1",
+										new double[] {0.0, 0.0},
+										"",
+										new int[] {0, 0},
+										new int[] {0, 0},
+										new int[] {0, 0},
+										""
+								),
+								ForageQuantity.ABUNDANT,
+								ForageQuality.EXCELLENT,
+								"never",
+								"notes"
+						)
+				)
```
better create variables forageWikiItem, customMarker and forageMapItem before line 222. then the when-Statement will be clearer

Phaete replied:
I don't quite understand what the problem is here. Everything is clearly separated with indentations making it clear that the Optional-Object I am building here is the supposed return value for the when-call. 
Why do I need to store everything in a variable before I am allowed to use it?

sieberss replied:
The problem is that one has to inspect intensively if there are differences between this object and the one you expect to be returned. And if so, are they intented or typos? ("Do not repeat yourself!")

When you put a variable "item" in and expect the same variable as result, it would be much easier to see the relevant part of the test.

sieberss reviewed (commented):
Much better now, but readability can be even more improved.

What about my code suggestiion?

Phaete commented:
We clearly have two different coding styles.

What is the metric you measure "readability" with?

Phaete committed:
Change all JUnits tests in services to create only a single service object instead of a new one each test method

Phaete committed:
Refactor ForageMapItemServiceTest as per reviewers demands

Phaete committed:
Fix tests in ForageMapItemServiceTest

sieberss commented:
> We clearly have two different coding styles.

Seems so.

> What is the metric you measure "readability" with?

The time I need to understand what's going on.

Maybe you also will wish better readability when you have to review code more complex than that of our co-students.

Phaete commented:
Since you are still not approving this PR, what other problems do you have with my code?

sieberss commented on backend/src/test/java/com/phaete/backend/forage/service/ForageMapItemServiceTest.java:
```diff
@@ -16,67 +16,69 @@ class ForageMapItemServiceTest {
-				"notes"
-		);
-
 		when(forageMapItemRepository.save(any(ForageMapItem.class))).thenReturn(
```
Now as your code has become clearer, I see that you use "any" here.
That is not necessary as you are mocking your IdService and can use a mocked id.

sieberss reviewed (approved):
I think this was hard work for both of us.

Now I can see that your code seems to work right. Before your changes I was not able to.
